### DISCUSSION TITLE
Insert node in the xml file directly

### DISCUSSION THREAD

**Discussion by mitien** [renggli/dart-xml#129] [Q&A] [answered]
Is there any way to insert nodes directly in xml file without recreating full xml document with builder?

```xml
<rootNode>
...
  <mainNode>
...
    <randomNode>111</randomNode>
    <randomNode2>1</randomNode2>
    <randomNode3><subNode1><dataId>1</dataId></subNode1></randomNode3>
    <item name="Text1"><dataValue>1</dataValue></item >
    <item name="Text2"><dataValue>2</dataValue></item >
    <item name="Text3"><dataValue>3</dataValue></item >

    >>>>>> insert here ( example <item name="Text4"><dataValue>4</dataValue></item > ) <<<<<<

    <subNode2><child>text text</child></subNode2>
...
  <mainNode>
...
</rootNode>
```

**Comment by renggli** [comment, marked as answer]
Sure, something along those lines should work:

```dart
final document = XmlDocument.parse(input);
final mainNode = document.findAllElements('mainNode').first;
final lastItemNode = mainNode.findAllElements('item').last;
final lastItemIndex = mainNode.children.indexOf(lastItemNode);
final builder = XmlBuilder();
builder.element('item',
     attributes: {'name': 'Text4'},
     nest: () => builder.element('dataValue', nest: 4));
mainNode.children.insert(lastItemIndex + 1, builder.buildFragment());
final output = document.toXmlString();
```

**Reply by mitien** [reply]
Thanks a lot is helps.

One more question if you don\`t mind.

Both  `builder.buildFragment()`  and `builder.buildDocument()` are one-timers. Right?

```
//print children
print(builder.buildFragment().children);  
//print []
print(builder.buildFragment().children);  
```

**Comment by renggli** [comment]
Correct, see [builder.dart#L252](https://github.com/renggli/dart-xml/blob/fc6593a612eab992fe9a84f88ab675ff1bd08679/lib/src/xml/builder.dart#L250).

If you want multiple instances you can copy the built tree:

```dart
final item = builder.buildFragment();
final copy = item.copy();
```

**Reply by mitien** [reply]
Ok. Got it.

But new question appear.  Please correct me if I`m wrong.

If use event driven API strategy will be the same?
1) find out index of element
2) find that item in the main tree
3) update/remove/insert new.

if need to do the same with multiple nodes - rinse and repeat.


Thanks in advance

**Reply by renggli** [reply]
With the event driven API the solution will be very different: you need to identify the place when to insert/transform/modify something, perform this transformation by editing the generated events, and then serializing it back to a string.

**Reply by mitien** [reply]
Thanks. Understood.
